sycl/doc/Array_Kernel_Parameters.md

```diff
@@ -0,0 +1,435 @@
+This includes scalars, pointers, and standard-layout structs.
+Certain SYCL struct types that are not standard-layout,
+such as Accessors and Samplers, are treated specially.
+The arguments to their init functions are passed as separate parameters
+and used within the kernel caller function to initialize Accessors/Samplers
+on the device by calling their init functions using the received arguments.
+
+There is one other aspect of code generation. An �integration header�
+is generated for use during host compilation.
+This header file contains entries for each kernel.
+Among the items it defines is a table of sizes and offsets of the
+kernel parameters.
+For the source example above the integration header contains the
+following snippet:
+
+```C++
+// array representing signatures of all kernels defined in the
+// corresponding source
+static constexpr
+const kernel_param_desc_t kernel_signatures[] = {
+  //--- _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE19->18clES2_E6Worker
+  { kernel_param_kind_t::kind_accessor, 4062, 0 },
+  { kernel_param_kind_t::kind_std_layout, 4, 32 },
+  { kernel_param_kind_t::kind_std_layout, 4, 36 },
+};
+```
+
+Each entry in the kernel_signatures table contains three values:
+1) an encoding of the type of capture object member
+2) a field that encodes additional properties, and
+3) an offset within a block of memory where the value of that
+4) kernel argument is placed.
+
+The previous sections described how kernel arguments are handled today.
+The next three sections describe support for arrays.
+
+<h3>Fix 1: Kernel Arguments that are Standard-Layout Arrays</h3>
+
+As described earlier, each variable captured by a lambda that comprises a
+SYCL kernel becomes a parameter of the kernel caller function.
+For arrays, simply allowing them through would result in a
+function parameter of array type. This is not supported in C++.
+Therefore, the array needing capture is wrapped in a struct for
+the purposes of passing to the device. Once received on the device
+within its wrapper, the array is copied into the local capture object.
+All references to the array within the kernel body are directed to
+the non-wrapped array which is a member of the local capture object.
+
+<h4>Source code fragment:</h4>
+
+```C++
+  int array[100];
+  auto outBuf = buffer<int, 1>(&output[0], num_items);
+
+  myQueue.submit([&](handler &cgh) {
+    auto outAcc = outBuf.get_access<access::mode::write>(cgh);
+    cgh.parallel_for<class Worker>(num_items, [=](cl::sycl::id<1> index) {
+      outAcc[index] = array[index.get(0)];
+    });
+  });
+```
+
+<h4>Integration header produced:</h4>
+
+```C++
+static constexpr
+const kernel_param_desc_t kernel_signatures[] = {
+  //--- _ZTSZZ4mainENKUlRN2cl4sycl7handlerEE16->18clES2_E6Worker
+  { kernel_param_kind_t::kind_accessor, 4062, 0 },
+  { kernel_param_kind_t::kind_std_layout, 400, 32 },
+};
+```
+
+<h4>The changes to device code made to support this extension, in pseudo-code:</h4>
+
```

Comment:
These changes are made by the device compiler, right? Please add this not, otherwise not quite clear.
Also 'changes' -> 'transformations'?